src/main/java/org/jenkinsci/plugins/docker/workflow/WithContainerStep.java

```diff
@@ -141,9 +142,18 @@ public Execution() {
             workspace.mkdirs(); // otherwise it may be owned by root when created for -v
             String ws = getPath(workspace);
             toolName = step.toolName;
-            DockerClient dockerClient = launcher.isUnix()
-                ? new DockerClient(launcher, node, toolName)
-                : new WindowsDockerClient(launcher, node, toolName);
+            DockerClient dockerClient;
+            
+            if (launcher.isUnix()) {
+                dockerClient = new DockerClient(launcher, node, toolName);
+            }
+            else {
+                dockerClient = new WindowsDockerClient(launcher, node, toolName);
+                String os = dockerClient.inspect(new EnvVars(), step.image, ".Os");
+                if (os != null && os.equals("linux")) {
```

Comment:
Should it be actually inverted? There is a lot of Unix operating systems being containerized, but Windows Client is for Windows only?